system-test/read-rows.ts

```diff
@@ -123,97 +130,93 @@ describe('Bigtable/Table', () => {
+      server.shutdown(() => {});
     });
 
     tests.forEach(test => {
-      it(test.name, () => {
-        responses = test.responses;
-        TABLE.maxRetries = test.max_retries;
-        TABLE.createReadStream(test.createReadStream_options)
-          .on('data', row => rowKeysRead[rowKeysRead.length - 1].push(row.id))
-          .on('end', () => (endCalled = true))
-          .on('error', err => (error = err as ServiceError));
-        clock.runAll();
-
-        if (test.error) {
-          assert(!endCalled, ".on('end') should not have been invoked");
-          assert.strictEqual(error!.code, test.error);
-        } else {
-          assert(endCalled, ".on('end') shoud have been invoked");
-          assert.ifError(error);
+      it(test.name, done => {
+        // These variables store request/response data capturing data sent
+        // and received when using readRows with retries. This data is evaluated
+        // in checkResults at the end of the test for correctness.
+        const requestedOptions: google.bigtable.v2.IRowSet[] = [];
+        const responses = test.responses;
+        const rowKeysRead: string[][] = [];
+        let endCalled = false;
+        let error: ServiceError | null = null;
+        function checkResults() {
+          if (test.error) {
+            assert(!endCalled, ".on('end') should not have been invoked");
+            assert.strictEqual(error!.code, test.error);
+          } else {
+            assert(endCalled, ".on('end') should have been invoked");
+            assert.ifError(error);
+          }
+          assert.deepStrictEqual(rowKeysRead, test.row_keys_read);
+          assert.strictEqual(
+            responses.length,
+            0,
+            'not all the responses were used'
+          );
+          assert.deepStrictEqual(requestedOptions, test.request_options);
+          done();
         }
-        assert.deepStrictEqual(rowKeysRead, test.row_keys_read);
-        assert.strictEqual(
-          responses.length,
-          0,
-          'not all the responses were used'
-        );
-        assert.deepStrictEqual(requestedOptions, test.request_options);
+
+        table.maxRetries = test.max_retries;
+        service.setService({
+          ReadRows: (
+            stream: ServerWritableStream<
+              protos.google.bigtable.v2.IReadRowsRequest,
+              protos.google.bigtable.v2.IReadRowsResponse
+            >
+          ) => {
+            const response = responses!.shift();
+            assert(response);
+            rowKeysRead.push([]);
+            requestedOptions.push(getRequestOptions(stream.request));
+            if (response.row_keys) {
+              stream.write({
+                chunks: response.row_keys.map(rowResponseFromServer),
+              });
+            }
+            if (response.end_with_error) {
+              // eslint-disable-next-line @typescript-eslint/no-explicit-any
+              const error: any = new Error();
```

Comment:
Is this an error coming from the server? If so is it a GoogleError not an any?

Reply:
Yup. Changed to GoogleError.